src/test/java/seedu/recipe/testutil/TypicalRecipes.java

```diff
@@ -232,6 +199,39 @@ public class TypicalRecipes {
                             + "if you want heat."),
                     new Step("Assemble together and enjoy!"))).build();
 
+    public static final String CACIO_STRING = String.format(
+            "%s;\nPortion: %s;\nDuration: %s;\nTags: %s;\n"
+                    + "Ingredients: %s;\nSteps: %s",
+            CACIO_NAME, CACIO_PORTION, CACIO_DURATION, (
+            (Supplier<String>) () -> {
+                StringBuilder out = new StringBuilder();
+                ArrayList<Tag> tags = new ArrayList<>(CACIO_TAGS);
+                tags.sort(Comparator.comparing(tag -> tag.tagName));
+                for (Tag tag : tags) {
+                    out.append(tag.toString());
+                }
+                return out.toString();
+            }).get(), (
+            (Supplier<String>) () -> {
+                StringBuilder out = new StringBuilder();
+                for (IngredientBuilder i : CACIO_INGREDIENTS) {
+                    out.append(i.toString())
+                            .append(",\n");
+                }
+                return out.toString();
+            }).get(), (
+            (Supplier<String>) () -> {
+                StringBuilder out = new StringBuilder();
+                for (int i = 0; i < CACIO_STEPS.size(); i++) {
+                    out.append(i + 1)
+                            .append(". ")
+                            .append(CACIO_STEPS.get(i).toString())
+                            .append(",\n");
+                }
+                return out.toString();
+            }).get()
+    );
```

Comment:
the conversion of Tag/Ingredient/Step collections to String could be abstracted out into each of those classes. `TypicalRecipes` should not be responsible for implementing this behaviour,,, also this violates S L A P 👋

e.g. you could have something like `Tag::collectionToString` that does that conversion for you